src/firefly/js/visualize/projection/WavelengthHeaderParser.js

```diff
@@ -18,26 +18,186 @@ import {isDefined} from '../../util/WebUtil.js';
  * @prop {TableModel} table - the table with wavelength algorythm
  */
 
+
 /**
  *
  * @param {FitsHeader} header
  * @param {String} altWcs
  * @param {FitsHeader} zeroHeader
  * @param {Array.<WavelengthTabRelatedData>} wlTableRelatedAry
- * @return {*}
+ * @return {SpectralWCSData|undefined}
  */
 export function parseWavelengthHeaderInfo(header, altWcs = '', zeroHeader, wlTableRelatedAry) {
     const parse = makeDoubleHeaderParse(header, zeroHeader, altWcs);
-    const which = altWcs ? '1' : getWCSAXES(parse);
-    const whatType = parse.getValue(`CTYPE${which}${altWcs}`, ' ');
-    const mijMatrixKeyRoot = getPC_ijKey(parse, which);
-    if (!mijMatrixKeyRoot) return; //When both PC i_j and CD i_j are present, we don't show the wavelength
-    const table = findWLTableToMatch(parse, altWcs, which, wlTableRelatedAry);
-    if (whatType.toUpperCase().startsWith('WAVE')) {
-        return calculateWavelengthParams(parse, altWcs, which, mijMatrixKeyRoot, table);
-    } else if (whatType.toUpperCase() === VRAD) {
-        return calculateVradParams(parse, altWcs, which, mijMatrixKeyRoot, table);
+
+    // identify spectral coordinates and the subset that requires table lookup (-TAB) algorithm
+    const {iCtypeSupported, iCtypeTab} = findSupportedCTYPEs(parse, altWcs, [WAVE, VRAD]); // array of i, for which CTYPEi=XXXX-TAB
+
+    if (iCtypeSupported.length === 0) return;
+
+    const spectralCoords = [];
+    const failWarnings = []
+
+    // fields that belong to WCS, not to a specific spectral coordinate
+    const {nAxis, N, r_j, restWAV} = calculateSharedParams(parse, altWcs);
+    if (N < 0) return;  // Dimension value is not available, must be NAXIS or WCSAXES
+
+    // loop through spectral coordinates, assemble per-coordinate parameters
+    iCtypeSupported.forEach((which) => {
+        const {ctype, cname, units, crpix, crval, cdelt} = getCoreSpectralHeaders(parse, altWcs, which);
+
+        const mijMatrixKeyRoot = getPC_ijKey(parse, which);
+        if (!mijMatrixKeyRoot) {
+            failWarnings.push(`${ctype}: both PC i_j and CD i_j are present`);
+            return;
+        }
+
+        let pc_3j = parse.getDoubleAry(`${mijMatrixKeyRoot}${which}_`, altWcs, 1, N, undefined);
+        if (!pc_3j && mijMatrixKeyRoot === 'CD') {
+            // CD matrix values default to 0.0 - the spectral coordinate value is constant
+            failWarnings.push(`CD3_i${altWcs} is not defined: ${ctype} value is constant`);
+        }
+
+        //check if any value is not defined, use default
+        pc_3j = applyDefaultValues(pc_3j, mijMatrixKeyRoot, which, N);
+
+        const {algorithm, coordType} = getAlgorithmAndType(ctype);
+
+        const canDoPlaneCalc = nAxis === 3 && pc_3j?.length === 3 && pc_3j[0] === 0 && pc_3j[1] === 0 &&
+            allGoodValues(crpix, crval, cdelt && nAxis === 3);
```

Comment:
In this context pc_3j is the PC/CD transformation matrix row that corresponds to the spectral coordinate.

We assume that the spectral coordinate value should be displayed per plane  (not per pixel) when there are 3 image coordinates and the spectral coordinate has no dependency on the first two image coordinates.

Is it correct? Should I also check that the value corresponding to the 3-rd image axis is the only non-zero?